src/AutoSaveStatus/AutoSaveStatusProvider.tsx

```diff
@@ -0,0 +1,62 @@
+import React, { PropsWithChildren, useCallback, useEffect, useState } from "react";
+
+export enum AutoSaveStatus {
+  IDLE = "idle",
+  SAVING = "saving",
+  DONE = "done",
+  ERROR = "error",
+}
+
+export interface AutoSaveStatusContextType {
+  status: AutoSaveStatus;
+  /** Resets status to IDLE, particularly useful if "Error" or "Done" is stale */
+  resetStatus: VoidFunction;
+  errors: unknown[];
+  /** Notifies AutoSaveContext that a request is in-flight */
+  triggerAutoSave: VoidFunction;
+  /** Notifies AutoSaveContext that a request has settled, optionally taking an error */
+  resolveAutoSave: (error?: unknown) => void;
+}
+
+export const AutoSaveStatusContext = React.createContext<AutoSaveStatusContextType>({
+  status: AutoSaveStatus.IDLE,
+  resetStatus() {},
+  errors: [],
+  triggerAutoSave() {},
+  resolveAutoSave() {},
+});
+
+export function AutoSaveStatusProvider({ children }: PropsWithChildren<{}>) {
+  const [status, setStatus] = useState(AutoSaveStatus.IDLE);
+  const [errors, setErrors] = useState<unknown[]>([]);
+  const [inFlight, setInFlight] = useState(0);
+
+  /** Handles setting Status */
+  useEffect(() => {
+    if (inFlight > 0) return setStatus(AutoSaveStatus.SAVING);
+    if (status === AutoSaveStatus.IDLE) return;
+    if (errors.length) return setStatus(AutoSaveStatus.ERROR);
+    return setStatus(AutoSaveStatus.DONE);
+  }, [errors.length, inFlight, status]);
+
+  const triggerAutoSave = useCallback(() => {
+    setInFlight((c) => c + 1);
+    setErrors([]);
+  }, []);
+
+  const resolveAutoSave = useCallback((error?: unknown) => {
+    setInFlight((c) => Math.max(0, c - 1));
+    if (error) setErrors((errs) => errs.concat(error));
+  }, []);
+
+  const resetStatus = useCallback(() => {
+    setStatus(AutoSaveStatus.IDLE);
+    setErrors([]);
+  }, []);
+
+  return (
+    <AutoSaveStatusContext.Provider value={{ status, resetStatus, errors, triggerAutoSave, resolveAutoSave }}>
```

Comment:
**question:** Should we memoize this object? Since it is not a deep equality check for the Context's value then the value will be a new object every time this provider re-renders. Each of the properties in the object would retain a stable identity, but the object as a whole will look new, triggering an unnecessary re-render.

Reply:
We can but I've never seen it done. `const { state } = useAutoSaveStatus();`, because of the destructure, would be stable for anything that depends on `state`.

Reply:
Okay cool. Fine with me. I was under the impression that passing a non-memoized `value` would cause unwanted re-renders. If it is unnecessary, then great!